server/backup_binlog_grpc.go

```diff
@@ -0,0 +1,189 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/cybozu-go/log"
+	"github.com/cybozu-go/moco"
+	"github.com/cybozu-go/moco-agent/metrics"
+	"github.com/cybozu-go/moco-agent/server/proto"
+	"github.com/cybozu-go/well"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// NewBackupBinlogService creates a new BackupBinlogServiceServer
+func NewBackupBinlogService(agent *Agent) proto.BackupBinlogServiceServer {
+	return &backupBinlogService{
+		agent: agent,
+	}
+}
+
+type backupBinlogService struct {
+	proto.UnimplementedBackupBinlogServiceServer
+	agent *Agent
+}
+
+// FlushAndBackupBinaryLogs executes "FLUSH BINARY LOGS;"
+// and upload it to the object storage, then delete it
+func (s *backupBinlogService) FlushAndBackupBinlog(ctx context.Context, req *proto.FlushAndBackupBinlogRequest) (*proto.FlushAndBackupBinlogResponse, error) {
+	if req.Token != s.agent.token {
+		return nil, status.Error(codes.Unauthenticated, "invalid token")
+	}
+
+	if !s.agent.sem.TryAcquire(1) {
+		return nil, status.Error(codes.ResourceExhausted, "another request is under processing")
+	}
+
+	sess := session.Must(session.NewSession(&aws.Config{
+		Region:           aws.String(req.BucketRegion),
+		Endpoint:         aws.String(fmt.Sprintf("%s:%d", req.BucketHost, req.BucketPort)),
+		DisableSSL:       aws.Bool(true),
+		S3ForcePathStyle: aws.Bool(true),
+		Credentials:      credentials.NewStaticCredentials(req.AccessKeyId, req.SecretAccessKey, ""),
+	}))
+
+	// prevent re-execution of the same request
+	_, err := s3.New(sess).HeadObject(&s3.HeadObjectInput{
+		Bucket: aws.String(req.BucketName),
+		Key:    aws.String(req.BackupId),
+	})
+	if err == nil {
+		s.agent.sem.Release(1)
+		return nil, status.Errorf(codes.InvalidArgument, "the requested backup has already completed: BackupId=%s", req.BackupId)
+	}
+	awsErr, ok := err.(awserr.RequestFailure)
+	if !ok {
+		s.agent.sem.Release(1)
+		log.Error("unknown response from object storage", map[string]interface{}{
+			log.FnError: err,
+		})
+		return nil, status.Errorf(codes.Internal, "unknown response from object storage: err=%+v", err)
+	}
+	if awsErr.StatusCode() != http.StatusNotFound {
+		s.agent.sem.Release(1)
+		log.Error("failed to get objects", map[string]interface{}{
+			log.FnError: err,
+		})
+		return nil, status.Errorf(codes.Internal, "failed to get object: err=%+v", err)
+	}
+
+	// TODO: change user
```

Comment:
Could you add an issue? I'm afraid to forget to update here.

Reply:
Done.
https://github.com/cybozu-go/moco/issues/208